helix-core/src/main/java/org/apache/helix/controller/rebalancer/MaintenanceRebalancer.java

```diff
@@ -55,12 +59,19 @@ public IdealState computeNewIdealState(String resourceName, IdealState currentId
 
     // One principal is to prohibit DROP -> OFFLINE and OFFLINE -> DROP state transitions.
     // Derived preference list from current state with state priority
+    StateModelDefinition stateModelDef = clusterData.getStateModelDef(currentIdealState.getStateModelDefRef());
+
     for (Partition partition : currentStateMap.keySet()) {
       Map<String, String> stateMap = currentStateMap.get(partition);
       List<String> preferenceList = new ArrayList<>(stateMap.keySet());
+
+      // This sorting preserves the ordering of current state hosts in the order of current IS pref list
       Collections.sort(preferenceList, new PreferenceListNodeComparator(stateMap,
-          clusterData.getStateModelDef(currentIdealState.getStateModelDefRef()),
-          Collections.<String>emptyList()));
+          stateModelDef, currentIdealState.getPreferenceList(partition.getPartitionName())));
+
+      // This sorts the current state hosts based on the priority
+      preferenceList.sort(new StatePriorityComparator(stateMap, stateModelDef));
```

Comment:
This will not help, right? It was the problem the order of two followers are not deterministic.

Reply:
So this is getting applied when we have already sorted the current state hosts in the preference-list order. (line: 70)
```
pref-list: [A, B, C]
current-state: {
    A: FOLLOWER,
    B: LEADER,
    C: FOLLOWER
}
```
Lets say `newPrefList = new ArrayList<>(current-state.keySet())` => `[C, B, A]`
```
Sort 1: Sort based on preference-list order:
newPrefList = [C, B, A] => [A, B, C]

Sort 2: Sort based on state-priority order:
newPrefList = [A, B, C] => [B, A, C]
```
Here, A will be 2nd and C will be third always as both have same priority so original (pref-list) order will be maintained.

Reply:
May be adding comment with details will be useful.

Reply:
Yes, updated code with an example as well. Thanks Komal!